mlflow/types/utils.py

```diff
@@ -78,6 +87,117 @@ def _get_str_or_byte_type(data):
         return DataType.binary
 
 
+def _infer_datatype(data) -> DataType:
+    if DataType.is_boolean(data):
+        return DataType.boolean
+    if DataType.is_integer(data):
+        return DataType.integer
+    if DataType.is_long(data):
+        return DataType.long
+    if DataType.is_float(data):
+        return DataType.float
+    if DataType.is_double(data):
+        return DataType.double
+    if DataType.is_string(data):
+        return DataType.string
+    if DataType.is_binary(data):
+        return DataType.binary
+    if DataType.is_datetime(data):
+        return DataType.datetime
+
+
+def _update_object_properties(obj1: Object, obj2: Object) -> Object:
+    if obj1 == obj2:
+        return obj1
+    properties1 = {prop.name: prop for prop in obj1.properties}
+    properties2 = {prop.name: prop for prop in obj2.properties}
+    updated_properties = []
+    # For each property in the first element, if it doesn't appear
+    # later, we update required=False
+    for k in properties1.keys() - properties2.keys():
+        properties1[k].required = False
+        updated_properties.append(properties1[k])
+    # For common keys, property type should be the same
+    for k in properties1.keys() & properties2.keys():
+        updated_properties.append(_update_property_type(properties1[k], properties2[k]))
+    # For each property appears in the second elements, if it doesn't
+    # exist, we update and set required=False
+    for k in properties2.keys() - properties1.keys():
+        properties2[k].required = False
+        updated_properties.append(properties2[k])
+    return Object(properties=updated_properties)
+
+
+def _update_property_type(prop1: Property, prop2: Property) -> Property:
+    if prop1.name != prop2.name:
+        raise MlflowException("Can't update two properties with different names")
+    required = prop1.required and prop2.required
+    if isinstance(prop1.dtype, DataType) and isinstance(prop2.dtype, DataType):
+        if prop1.dtype == prop2.dtype:
+            return Property(name=prop1.name, dtype=prop1.dtype, required=required)
+        raise MlflowException(f"Properties are incompatible for {prop1.dtype} and {prop2.dtype}")
+
+    if isinstance(prop1.dtype, Object) and isinstance(prop2.dtype, Object):
+        return Property(name=prop1.name, dtype=_update_object_properties(prop1.dtype, prop2.dtype))
+
+    if isinstance(prop1.dtype, Array) and isinstance(prop2.dtype, Array):
+        if prop1.dtype.dtype == prop2.dtype.dtype:
+            return Property(name=prop1.name, dtype=prop1.dtype, required=required)
+        if isinstance(prop1.dtype.dtype, Object) and isinstance(prop2.dtype.dtype, Object):
+            object = _update_object_properties(prop1.dtype.dtype, prop2.dtype.dtype)
```

Comment:
```suggestion
            obj = _update_object_properties(prop1.dtype.dtype, prop2.dtype.dtype)
```

`object` sounds like the built-in `object`.